### PR TITLE
Add DS data validation for existing domains and update failure message

### PR DESCRIPTION
**Pull request by ptkach**
As agreed, I'm making existing domains DS validation failure more explicit and consistent.

- - -
This change is [Reviewable](https://reviewable.io/reviews/google/nomulus/1897)

### CODE REVIEW

File: core/src/test/java/google/registry/flows/domain/DomainUpdateFlowTest.java [hunk 1]
```diff
@@ -1047,8 +1063,10 @@
   @Test
   void testFailure_secDnsTooManyDsRecords() throws Exception {
     ImmutableSet.Builder<DomainDsData> builder = new ImmutableSet.Builder<>();
+    byte[] b = new byte[32];
+    new Random().nextBytes(b);
```
[the remark this answers is not in the document]
Comment (ptkach) [reply core/src/test/java/google/registry/flows/domain/DomainUpdateFlowTest.java:1067]: This warning is not applicable to this line as bytes itself do not affect the test, but rather length is verified, thus byte array instance is more efficient.

**Review by weiminyu — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @ptkach)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 324 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFnNLCp_FRVcGEiCv:-NLMFnNMBcGMhk9Yms5F:b-h06odm) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L324)):*
> ```Java
>     try {
>       validateDsData(dsData);
>     } catch (InvalidDsRecordException e) {
> ```

Why not change at the throwing site? 



_Code quote:_
```Java
InvalidDsRecordException
```

**Review by weiminyu — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @ptkach)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 189 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMGG15C3sVLN6FOthx:-NLMGG15C3sVLN6FOthy:b-3rx35a) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L189)):*
> ```Java
> 
>   private static final String EXISTING_DS_DATA_INVALID =
>       "Existing DS Data is invalid and has to be corrected or removed. ";
> ```

How about:

Invalid DS Data must be corrected or removed. Hint: you can use one EPP request to fix DS Data and perform other updates.



_Code quote:_
```Java
Existing DS Data is invalid and has to be corrected or removed.
```

**Review by jianglai — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @github-code-scanning[bot], @ptkach, and @weiminyu)

___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 180 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFRhM5tUpB_0mEOAh:-NLMFRhM5tUpB_0mEOAi:b-oxfpal) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L180)):*
> ```Java
>     Domain existingDomain = loadAndVerifyExistence(Domain.class, targetId, now);
>     verifyUpdateAllowed(command, existingDomain, now);
>     validateExistingDsData(existingDomain.getDsData());
> ```

It seems a bit counter-intuitive to validate the existing DS data here, and then to validate the new data immediately below (and its duplicate work when the DS data is not updated).

I understand that the goal is to provide a more descriptive error message, but I think we can probably do better by wrapping the statements in `validateNewState` in a try loop that re-throws the error and add a disclaimer in the error message that the invalid data *may* be already existing. You can also cover other potential validation errors (like the one thrown by `validateRequiredContactsPresent`) this way, as opposed to only explicitly checking existing DS data violations.



_Code quote:_
```Java
validateExistingDsData(existingDomain.getDsData());
```

**Review by ptkach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @weiminyu)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 189 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMGG15C3sVLN6FOthx:-NLMP3OR92Cym957AL7v:bhf1714) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L189)):*
<details><summary><i>Previously, weiminyu (Weimin Yu) wrote…</i></summary><blockquote>

How about:

Invalid DS Data must be corrected or removed. Hint: you can use one EPP request to fix DS Data and perform other updates.
</blockquote></details>

I will update, thanks
___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 324 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFnNLCp_FRVcGEiCv:-NLMPF3H7nvUpQgQIqKJ:b-brgnsd) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L324)):*
<details><summary><i>Previously, weiminyu (Weimin Yu) wrote…</i></summary><blockquote>

Why not change at the throwing site? 
</blockquote></details>

That'd make validateDsData method more complex as it'd need to accept parameter (likely boolean existing vs new DsData) unrelated to it's purpose and construct messages differently. I believe decorator pattern keeps things cleaner in this case. Thoughts? 
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 180 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFRhM5tUpB_0mEOAh:-NLMOCa53FIB8rl_FCmM:b-gx5cfy) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L180)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

It seems a bit counter-intuitive to validate the existing DS data here, and then to validate the new data immediately below (and its duplicate work when the DS data is not updated).

I understand that the goal is to provide a more descriptive error message, but I think we can probably do better by wrapping the statements in `validateNewState` in a try loop that re-throws the error and add a disclaimer in the error message that the invalid data *may* be already existing. You can also cover other potential validation errors (like the one thrown by `validateRequiredContactsPresent`) this way, as opposed to only explicitly checking existing DS data violations.
</blockquote></details>

I'm not exactly sure that would be the best way. Even though I agree it's a bit counter-intuitive, but it also allows us to exit sooner without performing unnecessary computations and from my perspective it better follows single responsibility principle. What do you think?
___
*[`core/src/test/java/google/registry/flows/domain/DomainUpdateFlowTest.java` line 1067 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NL8G2ME215reEzi7P6Z-r1-1067:-NLMOd8MAUPGo_zYEPGX:b9gccnx) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/test/java/google/registry/flows/domain/DomainUpdateFlowTest.java#L1067)):*
<details><summary><i>Previously, weiminyu (Weimin Yu) wrote…</i></summary><blockquote>

Then why do we need to use Random. Why not rename `b` as array_32_bytes and drop the call to Random?
</blockquote></details>

Good point, let me check.

**Review by ptkach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @weiminyu)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 189 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMGG15C3sVLN6FOthx:-NLMXbIl9D5RBdR0M_U7:b-bu91xm) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L189)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

I will update, thanks
</blockquote></details>

Updated
___
*[`core/src/test/java/google/registry/flows/domain/DomainUpdateFlowTest.java` line 1067 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NL8G2ME215reEzi7P6Z-r1-1067:-NLMXdiy4znMj9S8DzE0:b-bu91xm) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/test/java/google/registry/flows/domain/DomainUpdateFlowTest.java#L1067)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

Good point, let me check. 
</blockquote></details>

Updated

**Review by jianglai — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 1 of 3 files reviewed, 4 unresolved discussions (waiting on @github-code-scanning[bot], @ptkach, and @weiminyu)

___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 180 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFRhM5tUpB_0mEOAh:-NLMuZtpDQH7hlrNE4Lh:blf3ct5) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L180)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

I'm not exactly sure that would be the best way. Even though I agree it's a bit counter-intuitive, but it also allows us to exit sooner without performing unnecessary computations and from my perspective it better follows single responsibility principle. What do you think?
</blockquote></details>

Actually, I realized that we cannot validate existing data here because what if the update is exactly to correct the corrupt existing DS data? That would have prevented the update from going through...

**Review by weiminyu — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 1 of 3 files reviewed, 4 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @ptkach)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 324 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFnNLCp_FRVcGEiCv:-NLN-CziAK0x3iGGEMIa:bxtp41d) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L324)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

That'd make validateDsData method more complex as it'd need to accept parameter (likely boolean existing vs new DsData) unrelated to it's purpose and construct messages differently. I believe decorator pattern keeps things cleaner in this case. Thoughts? 
</blockquote></details>

I mean makding changes in DomainFlowUtils#validateDsData. You are rethrowing the same exception with a constant prelude, which is general enough to be pushed to that method.

The rethrow also truncates the stack trace. It is often better to include the caught exception as cause.
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 180 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFRhM5tUpB_0mEOAh:-NLN18WD-fq1YBlMdGKl:b-o5e6jy) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L180)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

Actually, I realized that we cannot validate existing data here because what if the update is exactly to correct the corrupt existing DS data? That would have prevented the update from going through...
</blockquote></details>

Agree. And we should have a test where a domain with bad DsData  is fixed.

**Review by jianglai — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @github-code-scanning[bot], @ptkach, and @weiminyu)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 189 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMGG15C3sVLN6FOthx:-NLWD_0fDC2YBEExFAAq:banax6p) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L189)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

Updated
</blockquote></details>

Personally I think the new language is more confusing than "existing DS data is invalid and has to be corrected or removed" is better than "invalid DS data must be corrected or removed" because if I were to update something other than DS records and get a message about "invalid DS data", I'd be very confused, whereas if the message tells me "existing DS data is invalid", I know the problem is not with my update request, which does not concern DS.

However more broadly, I feel that we're circling around the root issue here without addressing it. Since we only validate on the new data, there is always a chance that something existing might impact the update, not limited to just DS validation. Currently there are three validations in `validateNewState`, but we might add more in the future that makes some other existing data invalid. Just because invalid DS data was the one that got us attention doesn't mean we should just fix it. That was why I proposed early on that we do the following:

```
try {
  validateNewData(...);
} catch (Exception e) {
  String message = e.getMessage();
  throw new EppException(messsage + " Note: the invalid data may not be part of the update, check (and correct) existing data on the domain before attempting the update. You can also use one EPP request to fix the violation AND perform other updates attempted here", e);
}
```

We can not just change the error message at the site where they are thrown, because the same validation is used for both domain update and domain create, so we catch them and add the additional message to let the registrar know what *could* be the cause. This way we also don't have to worry about adding more validations in the future.
___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 324 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMFnNLCp_FRVcGEiCv:-NLWJqrIC7p1-Y-zU4V5:b-1gy414) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L324)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

Updated
</blockquote></details>

See my other comment, there is benefit in editing the message of the exception after it is thrown from `validateNewState`, namely it makes the code cleaner and it deals with any future validations that we might add. There is no reason why DS validation needs to be treated specially, and we already have three validations in `validateNewState`. It very well might be the case that all existing data pass the other two validations, but I think we should not rely on that, and should just make it so that the error message that eventually is surfaced to the client makes it obvious that the error *might* be in the existing data and tells them how to fix it.
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 312 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NLWEZ-ZEotrx5dxwEag:-NLWEZ-ZEotrx5dxwEah:b-g35wi4) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L312)):*
> ```Java
>   private void validateNewState(Domain newDomain, Domain existingDomain) throws EppException {
>     validateRequiredContactsPresent(newDomain.getRegistrant(), newDomain.getContacts());
>     try {
> ```

This try - catch -rethrow logic will be confusing to 



_Code quote:_
```Java
    try {
      validateDsData(newDomain.getDsData(), false);
    } catch (Exception newDsDataException) {
      try {
        validateDsData(existingDomain.getDsData(), true);
      } catch (Exception existingDsDatException) {
        throw existingDsDatException;
      }
      throw newDsDataException;
    }
```

**Review by weiminyu — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @ptkach)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 322 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NLbmDVA48RocJpP3F4N:-NLbmDVA48RocJpP3F4O:b-am5uzc) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L322)):*
> ```Java
> 
>   /** Check that the DS data that will be set on a domain is valid. */
>   static void validateDsData(Set<DomainDsData> dsData, boolean isExisting) throws EppException {
> ```

I didn't expect the code would look this. This is a bit awkward.

Maybe we should undo this. Sorry.

The validateNewState method seems a better place to handle errors.



_Code quote:_
```Java
 boolean isExisting
```
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 318 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NLbn0h_4NEjX8xvxYug:-NLbn0h_4NEjX8xvxYuh:b-c6cuhb) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L318)):*
> ```Java
>         validateDsData(existingDomain.getDsData(), true);
>       } catch (Exception existingDsDatException) {
>         throw existingDsDatException;
> ```

throw new EppException("Invalid data in domain. Please fix or remove it first. Hint ...", existingDataException);



_Code quote:_
```Java
throw existingDsDatException
```
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 320 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NLbnOR--QoZbe9O6rkC:-NLbnOR--QoZbe9O6rkD:b-ug47no) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L320)):*
> ```Java
>         throw existingDsDatException;
>       }
>       throw newDsDataException;
> ```

throw new EppException("Invalid data in request. Please fix ....", newDataException);



_Code quote:_
```Java
throw newDsDataException
```

**Review by jianglai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @github-code-scanning[bot] and @weiminyu)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 189 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMGG15C3sVLN6FOthx:-NMU2IlX-Cr6hit2zKuz:blzxksl) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L189)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

@weiminyu  @jianglai  EppException currently has private constructor and can't be instantiated, so unless I make it public there's no clean way for me to rethrow the error from `validateNewData` method. I'd very much rather not expose it as a public as it was clearly intended not to be as such. Without ability to rethrow EppException the only possible solutions are - have a new method to handle each exception separately and rethrow (my initial approach) or let it be taken care by `validateDsData` (my current approach). `validateNewData` shouldn't be responsible for exception construction logic.

 > We can not just change the error message at the site where they are thrown
- Can you elaborate? I have added a param `isExisting` specifically to change the message when we want it. Am I missing something? 

</blockquote></details>

What I meant was that you cannot change the error message without modifying the exception class.

Personally I'm fine with making `EppException`'s constructor public. The subtype matters less as it's `EppException` that is [caught](https://cs.opensource.google/nomulus/nomulus/+/master:core/src/main/java/google/registry/flows/EppController.java;l=69?q=%22catch%20(EppException%22&ss=nomulus%2Fnomulus) and converted to EPP response message.

My main concern is that (and I reiterate) in the current proposal, we are only treating existing DS data validation differently (because that's that brought our attention to the issue in the first place), whereas there are already 3 validations in `validateNewData`. We don't that there are no conflicting existing data for the other two validations.

So the way that I see it is we should address the problem in one of the two ways:

1) Add code to make sure that existing data validation failure would not prevent updates from going through, which is what this PR is trying to. If we do this, we should make sure that the logic we add is generic enough so that we don't have to worry about this issue again, not for the validations already there, or for future validations that we might add.

2) Given validation violation in existing data is rather rare, per Weimin, we could just go through all of them and fix them, either by hand or by informing the registrars. We make it clear to them that the existing data is bad and if they try to update domain without fixing it, the update would not go through. We can probably do the same audit on violations for the other two validations, which I suspect should turn up nothing as I think they were in place since inception, so no bad data would have gone through. We should also add a clear warning comment/javadoc in `validateNewState` telling future us that if we ever add more validations, we need to make sure existing data is compatible, lest it would cause issues when the affecting domains are updated.

Now, given what have been discussed extensively in this PR, it seems there really is no way to do option 1 cleanly. Should we instead entertain the idea of doing option 2 instead? After all, it is better to make sure your data maintains the sensible invariance that your code expects, than to wrestle with ourselves to write code that works with bad data. What do you think?

**Review by weiminyu — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @ptkach)

___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 317 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NM_18GA2JBEuFj8bS6e:-NM_18GA2JBEuFj8bS6f:b-m6ihxz) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L317)):*
> ```Java
>       try {
>         validateDsData(existingDomain.getDsData(), true);
>       } catch (Exception existingDsDatException) {
> ```

I'm surprised that this compiles. So the compiler can deduce that existingDsDataException is either unchecked or EppException.
But for readability, better narrow the type to EppException.

Also, should be  existingDsDat*a*Exception



_Code quote:_
```Java
Exception
```
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 320 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NLbnOR--QoZbe9O6rkC:-NM_-IsODld6w88GjV0K:b-7kudtl) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L320)):*
<details><summary><i>Previously, ptkach (Pavlo Tkach) wrote…</i></summary><blockquote>

EppException can't be instantiated as it is now and doesn't seem to be intended to. I'd rather not change that. 
</blockquote></details>

My main point is that our message should distinguish between bad data in database and bad data in user request. We shouldn't use the same message for both.

**Review by ptkach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1897)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @github-code-scanning[bot], @jianglai, and @weiminyu)

___
*[`core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java` line 189 at r1](https://reviewable.io/reviews/google/nomulus/1897#-NLMGG15C3sVLN6FOthx:-NM_LjDk29IqYJqHFYXi:b-swy55b) ([raw file](https://github.com/google/nomulus/blob/16c7e38a248b3c79bc8a2f5b9cdfa97eb5a24af0/core/src/main/java/google/registry/flows/domain/DomainFlowUtils.java#L189)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

What I meant was that you cannot change the error message without modifying the exception class.

Personally I'm fine with making `EppException`'s constructor public. The subtype matters less as it's `EppException` that is [caught](https://cs.opensource.google/nomulus/nomulus/+/master:core/src/main/java/google/registry/flows/EppController.java;l=69?q=%22catch%20(EppException%22&ss=nomulus%2Fnomulus) and converted to EPP response message.

My main concern is that (and I reiterate) in the current proposal, we are only treating existing DS data validation differently (because that's that brought our attention to the issue in the first place), whereas there are already 3 validations in `validateNewData`. We don't that there are no conflicting existing data for the other two validations.

So the way that I see it is we should address the problem in one of the two ways:

1) Add code to make sure that existing data validation failure would not prevent updates from going through, which is what this PR is trying to. If we do this, we should make sure that the logic we add is generic enough so that we don't have to worry about this issue again, not for the validations already there, or for future validations that we might add.

2) Given validation violation in existing data is rather rare, per Weimin, we could just go through all of them and fix them, either by hand or by informing the registrars. We make it clear to them that the existing data is bad and if they try to update domain without fixing it, the update would not go through. We can probably do the same audit on violations for the other two validations, which I suspect should turn up nothing as I think they were in place since inception, so no bad data would have gone through. We should also add a clear warning comment/javadoc in `validateNewState` telling future us that if we ever add more validations, we need to make sure existing data is compatible, lest it would cause issues when the affecting domains are updated.

Now, given what have been discussed extensively in this PR, it seems there really is no way to do option 1 cleanly. Should we instead entertain the idea of doing option 2 instead? After all, it is better to make sure your data maintains the sensible invariance that your code expects, than to wrestle with ourselves to write code that works with bad data. What do you think?
</blockquote></details>

@weiminyu  I agree with @jianglai that it's not really worth fixing considering that we don't have so many invalid values and it's better to have consistent approach. I will abandon this PR and make another with comment update for our future selfs to make sure that existing data is valid before updating conditions in `validateNewData`.  I will also fetch the list of invalid domains and reach out to gtech to contact customers. 
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 317 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NM_18GA2JBEuFj8bS6e:-NM_MTMN2AWVtXT-Mu-7:b-nzmq3g) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L317)):*
<details><summary><i>Previously, weiminyu (Weimin Yu) wrote…</i></summary><blockquote>

I'm surprised that this compiles. So the compiler can deduce that existingDsDataException is either unchecked or EppException.
But for readability, better narrow the type to EppException.

Also, should be  existingDsDat*a*Exception
</blockquote></details>

See my prev comment, I will abandon this PR.
___
*[`core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java` line 320 at r3](https://reviewable.io/reviews/google/nomulus/1897#-NLbnOR--QoZbe9O6rkC:-NM_MZuF0T8bT7ut2rG2:b-ey9ilt) ([raw file](https://github.com/google/nomulus/blob/344fb5e43a8528e4e05def51e32367374038e48d/core/src/main/java/google/registry/flows/domain/DomainUpdateFlow.java#L320)):*
<details><summary><i>Previously, weiminyu (Weimin Yu) wrote…</i></summary><blockquote>

My main point is that our message should distinguish between bad data in database and bad data in user request. We shouldn't use the same message for both.
</blockquote></details>

We should not have invalid data. See my prev comment, I will abandon this PR.

**Pull request closed without merging**